channels/parallel/client/parallel_main.c

```diff
@@ -241,31 +284,52 @@ static void* parallel_thread_func(void* arg)
 	IRP* irp;
 	wMessage message;
 	PARALLEL_DEVICE* parallel = (PARALLEL_DEVICE*) arg;
+	WIN32ERROR error = CHANNEL_RC_OK;
 
 	while (1)
 	{
 		if (!MessageQueue_Wait(parallel->queue))
+		{
+			WLog_ERR(TAG, "MessageQueue_Wait failed!");
+			error = ERROR_INTERNAL_ERROR;
 			break;
+		}
 
 		if (!MessageQueue_Peek(parallel->queue, &message, TRUE))
+		{
+			WLog_ERR(TAG, "MessageQueue_Peek failed!");
+			error = ERROR_INTERNAL_ERROR;
 			break;
+		}
 
 		if (message.id == WMQ_QUIT)
 			break;
 
 		irp = (IRP*) message.wParam;
 
-		parallel_process_irp(parallel, irp);
+		if ((error = parallel_process_irp(parallel, irp)))
+		{
+			WLog_ERR(TAG, "parallel_process_irp failed with error %d!", error);
+			break;
+		}
 	}
+	if (error && parallel->rdpcontext)
+		setChannelError(parallel->rdpcontext, error, "parallel_thread_func reported an error");
 
+	ExitThread((DWORD)error);
```

Comment:
I wonder if we do need this ExitThread here. If we return the error directly this should be sufficient in my opinion.